File: client/types/vetVisit.ts
export type VisitType =
  | "checkup"
  | "sick_visit"
  | "surgery"
  | "emergency"
  | "dental"
  | "grooming"
  | "follow_up"
  | "other";

export interface VetVisit {
  id: string;
  petId: string;
  ownerId: string;
  visitDate: string;
  visitType: VisitType;
  reason: string;
  vetName: string | null;
  clinicName: string | null;
  diagnosis: string | null;
  treatment: string | null;
  cost: number | null;
  followUpDate: string | null;
  followUpNeeded: boolean;
  notes: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface VetVisitFormPayload {
  petId: string;
  visitDate: string;
  visitType: VisitType;
  reason: string;
  vetName?: string;
  clinicName?: string;
  diagnosis?: string;
  treatment?: string;
  cost?: number;
  followUpDate?: string;
  followUpNeeded: boolean;
  notes?: string;
}

interface RawVetVisit {
  id: string;
  pet_id: string;
  owner_id: string;
  visit_date: string;
  visit_type: VisitType;
  reason: string;
  vet_name: string | null;
  clinic_name: string | null;
  diagnosis: string | null;
  treatment: string | null;
  cost: number | null;
  follow_up_date: string | null;
  follow_up_needed: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export function mapVetVisit(raw: RawVetVisit): VetVisit {
  return {
    id: raw.id,
    petId: raw.pet_id,
    ownerId: raw.owner_id,
    visitDate: raw.visit_date,
    visitType: raw.visit_type,
    reason: raw.reason,
    vetName: raw.vet_name,
    clinicName: raw.clinic_name,
    diagnosis: raw.diagnosis,
    treatment: raw.treatment,
    cost: raw.cost,
    followUpDate: raw.follow_up_date,
    followUpNeeded: raw.follow_up_needed,
    notes: raw.notes,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export function toVetVisitPayload(payload: VetVisitFormPayload) {
  return {
    pet_id: payload.petId,
    visit_date: payload.visitDate,
    visit_type: payload.visitType,
    reason: payload.reason,
    vet_name: payload.vetName || null,
    clinic_name: payload.clinicName || null,
    diagnosis: payload.diagnosis || null,
    treatment: payload.treatment || null,
    cost: payload.cost ?? null,
    follow_up_date: payload.followUpDate || null,
    follow_up_needed: payload.followUpNeeded,
    notes: payload.notes || null,
  };
}

File: server/models/vet_visit.py
import uuid
from datetime import datetime, date, timezone
import enum

from sqlalchemy import String, DateTime, Date, ForeignKey, Enum as SAEnum, Float, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.database.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class VisitType(str, enum.Enum):
    CHECKUP = "checkup"
    SICK_VISIT = "sick_visit"
    SURGERY = "surgery"
    EMERGENCY = "emergency"
    DENTAL = "dental"
    GROOMING = "grooming"
    FOLLOW_UP = "follow_up"
    OTHER = "other"


class VetVisit(Base):
    __tablename__ = "vet_visits"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    pet_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("pets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    owner_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    visit_date: Mapped[date] = mapped_column(Date, nullable=False)
    visit_type: Mapped[VisitType] = mapped_column(
        SAEnum(VisitType, name="visit_type"), default=VisitType.CHECKUP, nullable=False
    )
    reason: Mapped[str] = mapped_column(String(200), nullable=False)
    vet_name: Mapped[str | None] = mapped_column(String(120), nullable=True)
    clinic_name: Mapped[str | None] = mapped_column(String(120), nullable=True)
    diagnosis: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    treatment: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    cost: Mapped[float | None] = mapped_column(Float, nullable=True)
    follow_up_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    follow_up_needed: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    notes: Mapped[str | None] = mapped_column(String(1000), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, onupdate=_utcnow, nullable=False
    )

    pet = relationship("Pet", backref="vet_visits")
